### PR TITLE
common/async: add primitives for structured concurrency with optional_yield

### PR DESCRIPTION
**Pull request by cbodley**
~~(based on feature branch wip-boost-asio-spawn from https://github.com/ceph/ceph/pull/55592)~~

adds `spawn_throttle` class for bounded concurrency with stackful coroutines from `boost::asio::spawn()`. this relies on new support for per-op cancellation to guarantee that the lifetime of child coroutines won't exceed the lifetime of their throttle, making it safe for children to access memory from their parent's stack

by taking `optional_yield` in the constructor, it transparently supports both synchronous execution (where optional_yield is empty) and asynchronous execution within a stackful coroutine (where optional_yield contains the parent's yield_context)

also adds a `max_concurrent_for_each()` algorithm (inspired by [seastar](https://docs.seastar.io/master/group__future-util.html#ga2681cbe7e3bc3fe21d5314644bedb474)) to hide the use of `spawn_throttle`

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by cbodley** [comment]
this mimics the `spawn_group` interface from https://github.com/ceph/ceph/pull/50005 for `boost::asio::co_spawn()` (c++20 stackless coroutines). in hindsight, i'd probably rename that to `co_spawn_group` so this `yield_group` could use the name `spawn_group` instead

i also plan to create an `optional_yield` version of the `co_throttle` class from https://github.com/ceph/ceph/pull/49720 for bounded concurrency, and i'll covert `RGWDeleteMultiObj` to use that as a proof of concept

**Comment by cbodley** [comment]
@cfsnyder i used this `spawn_throttle` class to simplify `RGWDeleteMultiObj`. we now get concurrency in the null_yield case (where rgw_beast_enable_async=false) and no longer need the timer for `wait_flush()`

**Comment by cbodley** [comment]
jenkins test make check

**Comment by cbodley** [comment]
sorry for the review spam, i had to change the base branch

File: src/rgw/rgw_op.cc [hunk 1]
```diff
@@ -6767,43 +6767,38 @@ void RGWDeleteMultiObj::handle_individual_object(const rgw_obj_key& o, optional_
...
-  multi_delete = static_cast<RGWMultiDelDelete *>(parser.find_first("Delete"));
+  auto multi_delete = static_cast<RGWMultiDelDelete *>(parser.find_first("Delete"));
   if (!multi_delete) {
     op_ret = -EINVAL;
-    goto error;
-  } else {
+    return;
+  }
+
 #define DELETE_MULTI_OBJ_MAX_NUM      1000
```
Comment (adamemerson) [inline src/rgw/rgw_op.cc:6804]: nit: Since you're in the area could we make this a constant and not a `#define`?

**Review by adamemerson — APPROVED** [review]

**Comment by cbodley** [comment]
thanks @adamemerson. i decided to move the rgw stuff into https://github.com/ceph/ceph/pull/57641 so we can merge this one without qa

File: src/rgw/rgw_op.cc (hunk 1 above)
Comment (cbodley) [reply src/rgw/rgw_op.cc]: addressed in https://github.com/ceph/ceph/pull/57641. also added client-facing error messages

**Comment by cbodley** [comment]
p.s. once https://github.com/ceph/ceph/pull/56499 merges, i'd also like to add an overload of `max_concurrent_for_each()` based on `co_throttle`

**Comment by cbodley** [comment]
https://jenkins.ceph.com/job/ceph-api/74511/
> ERROR: setUpClass (tasks.mgr.dashboard.test_health.HealthTest)

https://tracker.ceph.com/issues/47612

**Comment by cbodley** [comment]
jenkins test api

**Comment by cbodley** [comment]
jenkins test windows

**Pull request merged**